### PR TITLE
Remove user from group

### PR DESCRIPTION
**Pull request by benrudolph**
@dannyroberts @emord what do you think about this approach? i don't love adding this property to the group model, but i also wasn't a huge fan of having the caller be responsible for saving the user removed

https://manage.dimagi.com/default.asp?252367

### CODE REVIEW

File: corehq/pillows/groups_to_user.py [hunk 1]
```diff
@@ -67,16 +68,29 @@ def update_es_user_with_groups(group_doc, es_client=None):
...
 def stream_user_sources(user_ids):
-    q = {"filter": {"and": [{"terms": {"_id": user_ids}}]}}
-    for result in stream_es_query(es_index='users', q=q, fields=["__group_ids", "__group_names"]):
-        group_ids = result.get('fields', {}).get("__group_ids", [])
+    results = (
+        UserES()
+        .user_ids(user_ids)
+        .fields(['_id', '__group_ids', '__group_names'])
+        .scroll()
+    )
```
Comment (benrudolph) [inline corehq/pillows/groups_to_user.py]: note this query is _slightly_ different from the original. this one excludes deleted and inactive users, which i think is fine, but maybe for consistency sake we should keep them up to date?

File: corehq/pillows/groups_to_user.py [hunk 2]
```diff
@@ -67,16 +68,29 @@ def update_es_user_with_groups(group_doc, es_client=None):
             doc = {"__group_ids": list(user_source.group_ids), "__group_names": list(user_source.group_names)}
             es_client.update(USER_INDEX, ES_META['users'].type, user_source.user_id, body={"doc": doc})
 
+    for user_source in stream_user_sources(group_doc.get("removed_users", [])):
+        if group_doc["name"] in user_source.group_names or group_doc["_id"] in user_source.group_ids:
+            user_source.group_ids.remove(group_doc["_id"])
+            user_source.group_names.remove(group_doc["name"])
+            doc = {"__group_ids": list(user_source.group_ids), "__group_names": list(user_source.group_names)}
+            es_client.update(USER_INDEX, ES_META['users'].type, user_source.user_id, body={"doc": doc})
```
Comment (emord) [inline corehq/pillows/groups_to_user.py]: Should the user be removed from `group.removed_users` at this point?

**Review by emord — COMMENTED** [review]
Curious why you used the group to do this? Does the user not get saved when they are removed from a group?

Really wish users/groups were in a relational db...

File: corehq/pillows/groups_to_user.py (hunk 2 above)
Comment (benrudolph) [inline corehq/pillows/groups_to_user.py]: yes it should, good catch

**Comment by benrudolph** [comment]
> Curious why you used the group to do this? Does the user not get saved when they are removed from a group?
users do not get saved when we remove them from a group 😞 one solution is to always save when removing a user from a group and then also save the users.
> Really wish users/groups were in a relational db...
yes, yes indeed

File: corehq/pillows/groups_to_user.py (hunk 2 above)
Comment (benrudolph) [inline corehq/pillows/groups_to_user.py]: err sorry, i think it's functionally fine as is. if we want to remove the user from that property it'd require us to do a couch fetch which seems meh. i don't see that property getting too large

**Comment by dannyroberts** [comment]
Clever. 👍

**Pull request merged**